# Eniac Product Analyse


## 1. Import the dataset

In [ ]:
from google.colab import drive
import pandas as pd

drive.mount('/content/drive')

PATH = "/content/drive/MyDrive/Project Eniac/Data/"

Mounted at /content/drive


In [ ]:
products = pd.read_csv(PATH + "products.csv")

##2. Check the Dataset Dimensions

In [ ]:
products

,sku,name,desc,price,promo_price,in_stock,type
0,RAI0007,Silver Rain Design mStand Support,Aluminum support compatible with all MacBook,59.99,499.899,1,8696
1,APP0023,Apple Mac Keyboard Keypad Spanish,USB ultrathin keyboard Apple Mac Spanish.,59,589.996,0,13855401
2,APP0025,Mighty Mouse Apple Mouse for Mac,mouse Apple USB cable.,59,569.898,0,1387
3,APP0072,Apple Dock to USB Cable iPhone and iPod white,IPhone dock and USB Cable Apple iPod.,25,229.997,0,1230
4,KIN0007,Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM,2GB RAM Mac mini and iMac (2006/07) MacBook Pr...,34.99,31.99,1,1364
...,...,...,...,...,...,...,...
19321,BEL0376,Belkin Travel Support Apple Watch Black,compact and portable stand vertically or horiz...,29.99,269.903,1,12282
19322,THU0060,"Enroute Thule 14L Backpack MacBook 13 ""Black",Backpack with capacity of 14 liter compartment...,69.95,649.903,1,1392
19323,THU0061,"Enroute Thule 14L Backpack MacBook 13 ""Blue",Backpack with capacity of 14 liter compartment...,69.95,649.903,1,1392
19324,THU0062,"Enroute Thule 14L Backpack MacBook 13 ""Red",Backpack with capacity of 14 liter compartment...,69.95,649.903,0,1392


In [ ]:
products.shape

(19326, 7)

The dataset contains 19,326 rows and 7 columns.

In [ ]:
products.size

135282

The total number of values that the DataFrame has

In [ ]:
products.shape[0] * products.shape[1] == products.size

True

##3.  Explore Products

In [ ]:
products.head()

,sku,name,desc,price,promo_price,in_stock,type
0,RAI0007,Silver Rain Design mStand Support,Aluminum support compatible with all MacBook,59.99,499.899,1,8696
1,APP0023,Apple Mac Keyboard Keypad Spanish,USB ultrathin keyboard Apple Mac Spanish.,59,589.996,0,13855401
2,APP0025,Mighty Mouse Apple Mouse for Mac,mouse Apple USB cable.,59,569.898,0,1387
3,APP0072,Apple Dock to USB Cable iPhone and iPod white,IPhone dock and USB Cable Apple iPod.,25,229.997,0,1230
4,KIN0007,Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM,2GB RAM Mac mini and iMac (2006/07) MacBook Pr...,34.99,31.99,1,1364


The dataset contains seven columns describing each product: the product identifier (sku), product name (name), description (desc), regular price (price), promotional price (promo_price), stock availability (in_stock), and product type (type).

##4. Explore the Dataset Structure

In [ ]:
products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19326 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          19326 non-null  object
 1   name         19326 non-null  object
 2   desc         19319 non-null  object
 3   price        19280 non-null  object
 4   promo_price  19326 non-null  object
 5   in_stock     19326 non-null  int64 
 6   type         19276 non-null  object
dtypes: int64(1), object(6)
memory usage: 1.0+ MB


The dataset contains 7 columns and 19,326 records.

##5. Check Data Types

In [ ]:
products.dtypes

,0
sku,object
name,object
desc,object
price,object
promo_price,object
in_stock,int64
type,object


Most variables are stored as the object data type, while in_stock is stored as an integer (int64). The price and promo_price columns are also stored as object, although they represent prices. These columns will be examined further during the data cleaning stage to determine whether a data type conversion is required.

##6. Check Missing Values

In [ ]:
products.isna().sum()

,0
sku,0
name,0
desc,7
price,46
promo_price,0
in_stock,0
type,50


In [ ]:
(products.isna().mean() * 100).round(2)

,0
sku,0.00
name,0.00
desc,0.04
price,0.24
promo_price,0.00
in_stock,0.00
type,0.26


The dataset contains missing values in three columns: desc (7 missing values), price (46 missing values), and type (50 missing values)

##7. Check Duplicate Records

In [ ]:
products.duplicated().sum()

np.int64(8746)

In [ ]:
products[products.duplicated()].head()

,sku,name,desc,price,promo_price,in_stock,type
101,APP0390,"Apple MacBook Pro 133 ""Core i5 25GHz | 4GB RAM...",MacBook Pro laptop 133 inches (MD101Y / A).,1199,11.455.917,0,1282
102,APP0390,"Apple MacBook Pro 133 ""Core i5 25GHz | 4GB RAM...",MacBook Pro laptop 133 inches (MD101Y / A).,1199,11.455.917,0,1282
103,APP0390,"Apple MacBook Pro 133 ""Core i5 25GHz | 4GB RAM...",MacBook Pro laptop 133 inches (MD101Y / A).,1199,11.455.917,0,1282
104,APP0390,"Apple MacBook Pro 133 ""Core i5 25GHz | 4GB RAM...",MacBook Pro laptop 133 inches (MD101Y / A).,1199,11.455.917,0,1282
105,APP0390,"Apple MacBook Pro 133 ""Core i5 25GHz | 4GB RAM...",MacBook Pro laptop 133 inches (MD101Y / A).,1199,11.455.917,0,1282


The dataset contains 8,746 duplicate rows

##8. Check Unique Values

In [ ]:
products.nunique()

,0
sku,10579
name,10373
desc,7098
price,2690
promo_price,4614
in_stock,2
type,126


The dataset contains a large number of unique product identifiers (sku) and product names (name), indicating a diverse product catalog

##9. Generate Descriptive Statistics

In [ ]:
products.describe(include="all")

,sku,name,desc,price,promo_price,in_stock,type
count,19326,19326,19319,19280,19326,19326.000000,19276
unique,10579,10373,7098,2690,4614,NaN,126
top,PAC1034,"Apple iMac 27 ""Core i7 Retina 5K 4GHz | 32GB |...",IMac desktop computer 27 inch 8GB RAM 512GB Re...,3649,28.739.896,NaN,"5,74E+15"
freq,58,89,230,466,229,NaN,6294
mean,NaN,NaN,NaN,NaN,NaN,0.109593,NaN
std,NaN,NaN,NaN,NaN,NaN,0.312390,NaN
min,NaN,NaN,NaN,NaN,NaN,0.000000,NaN
25%,NaN,NaN,NaN,NaN,NaN,0.000000,NaN
50%,NaN,NaN,NaN,NaN,NaN,0.000000,NaN
75%,NaN,NaN,NaN,NaN,NaN,0.000000,NaN


These statistics provide a useful overview of the dataset and highlight aspects that will be addressed during the data cleaning stage.

# Data Cleaning Products Dataset

##1. Create a Copy of the Dataset

In [ ]:
products_df = products.copy()

##2. Remove Duplicate Records

In [ ]:
products_df = products_df.drop_duplicates()

In [ ]:
products_df.duplicated().sum()

np.int64(0)

All duplicate rows were successfully removed from the dataset

##3. Review Dataset Information

In [ ]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10580 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          10580 non-null  object
 1   name         10580 non-null  object
 2   desc         10573 non-null  object
 3   price        10534 non-null  object
 4   promo_price  10580 non-null  object
 5   in_stock     10580 non-null  int64 
 6   type         10530 non-null  object
dtypes: int64(1), object(6)
memory usage: 661.2+ KB


##4. Handle Missing Values

In [ ]:
(products.isna().mean() * 100).round(2)

,0
sku,0.00
name,0.00
desc,0.04
price,0.24
promo_price,0.00
in_stock,0.00
type,0.26


In [ ]:
products_df.isna().sum()

,0
sku,0
name,0
desc,7
price,46
promo_price,0
in_stock,0
type,50


All rows containing missing values were successfully removed

##5. Correct Data Types

In [ ]:
products_df['price'].str.count(r"\.").value_counts()

,count
price,
1,6931
0,3173
2,373


In [ ]:
# @title Default title text
mult_decimal_rows = (products_df["price"].str.count(r"\.") > 1).sum()

percent_corrupted = (100 * mult_decimal_rows / products_df.shape[0])

print(f"{percent_corrupted:.2f}% of the rows in the products dataset have multiple decimal points in the price column.")

3.56% of the rows in the products dataset have multiple decimal points in the price column.


The price column contains 373 values with multiple decimal points, representing 3.56% of the dataset. Since these values cannot be correctly converted into a numeric format and represent only a small proportion of the data, these rows will be removed before converting the column to a numeric data type.

In [ ]:
products_df = products_df[products_df["price"].str.count(r"\.") <= 1]

In [ ]:
products_df.shape

(10104, 7)

In [ ]:
products_df["price"]

,price
0,59.99
1,59
2,59
3,25
4,34.99
...,...
19321,29.99
19322,69.95
19323,69.95
19324,69.95


In [ ]:
products_df["price"] = pd.to_numeric(products_df["price"])

In [ ]:
products_df.dtypes

,0
sku,object
name,object
desc,object
price,float64
promo_price,object
in_stock,int64
type,object


After removing the invalid values, the price column was successfully converted from object to float64

In [ ]:
products_df['promo_price'].str.count(r"\.").value_counts()

,count
promo_price,
1,5699
2,4284
0,121


In [ ]:
products_df = products_df.drop(columns=["promo_price"])

The products dataset was successfully cleaned.

In [ ]:
products_df.to_csv(PATH + "products_clean.csv", index=False)